src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNet/AspNetCoreMvc2Integration.cs

```diff
@@ -0,0 +1,324 @@
+        /// <param name="opCode">The OpCode used in the original method call.</param>
+        /// <param name="mdToken">The mdToken of the original method call.</param>
+        [InterceptMethod(
+            CallerAssembly = AspnetMvcCore,
+            TargetAssembly = AspnetMvcCore,
+            TargetType = DiagnosticSource,
+            TargetSignatureTypes = new[] { ClrNames.Void, ClrNames.Ignore, "Microsoft.AspNetCore.Mvc.Abstractions.ActionDescriptor", "Microsoft.AspNetCore.Http.HttpContext", "Microsoft.AspNetCore.Routing.RouteData" },
+            TargetMinimumVersion = Major2,
+            TargetMaximumVersion = Major2)]
+        public static void BeforeAction(
+            object diagnosticSource,
+            object actionDescriptor,
+            object httpContext,
+            object routeData,
+            int opCode,
+            int mdToken)
+        {
+            if (!Tracer.Instance.Settings.IsIntegrationEnabled(IntegrationName))
+            {
+                // integration disabled
+                return;
+            }
+
+            string methodDef = $"{DiagnosticSource}.{nameof(BeforeAction)}(...)";
+            var integrationContext = AspNetAmbientContext.RetrieveFromHttpContext(httpContext);
+
+            if (integrationContext == null)
+            {
+                Log.Error($"Could not access {nameof(AspNetAmbientContext)} for {methodDef}.");
+            }
+            else
+            {
+                SetAspNetCoreMvcSpecificData(
+                    ambientContext: integrationContext,
+                    actionDescriptor: actionDescriptor,
+                    httpContext: httpContext);
+            }
+
+            Action<object, object, object, object> instrumentedMethod = null;
+
+            try
+            {
+                instrumentedMethod =
+                    MethodBuilder<Action<object, object, object, object>>
+                       .Start(Assembly.GetCallingAssembly(), mdToken, opCode)
+                       .WithConcreteTypeName(DiagnosticSource)
+                       .WithParameters(diagnosticSource, actionDescriptor, httpContext, routeData)
+                       .Build();
+            }
+            catch (Exception ex)
+            {
+                // profiled app will continue working as expected without this method
+                Log.ErrorException($"Error resolving {DiagnosticSource}.{nameof(BeforeAction)}(...)", ex);
+            }
+
+            try
+            {
+                // call the original method, catching and rethrowing any unhandled exceptions
+                instrumentedMethod?.Invoke(diagnosticSource, actionDescriptor, httpContext, routeData);
+            }
+            catch (Exception ex)
+            {
+                // profiled app will continue working as expected without this method
+                Log.Error($"Exception when calling {methodDef}.", ex);
```

Comment:
Should we rethrow the original exception here?

Reply:
No, this is a carry forward from the original implementation of this integration.
This is a diagnostic method which does not affect the behavior of user code.
I am okay with throwing, but it seems safer to just log.

Reply:
We shouldn't change application behavior. If the instrumented method throws an exception, we should let it bubble up as if tracing was not enabled.